Allow to simulate usage of multiple catalogs by one connection

aschoerk opened the pull request:
I try to create tests which handle objects normally stored in a sybase database with multiple catalogs.
To support this using an in-memory H2 I did the following extension.

If IGNORECATALOG is enabled, catalognames in front of schema names will be ignored. This can be used if multiple catalogs used by the same connections must be simulated. Caveat: If both catalogs contain schemas of the same name and if those schemas contain objects of the same name, this will lead to errors, when trying to manage, access or change these objects.
This setting can be appended to the database URL: ""jdbc:h2:test;IGNORE_CATALOG=TRUE""

Also see similar problems in stackoverflow, which could be solved by this:
https://stackoverflow.com/questions/25088998/can-you-create-multiple-catalogs-in-h2
https://stackoverflow.com/questions/45220891/sql-server-entity-in-h2-schema-not-found

aschoerk committed:
Merge pull request #1 from h2database/master

get changes

Commit:
Support ignoring of Catalog-Values

Commit:
SET IGNORE_CATALOGS added, extended doc.

katzyn commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -4861,7 +4861,7 @@ private String readIdentifierWithSchema2(String s) {
         s = readColumnIdentifier();
         if (currentTokenType == DOT) {
-            if (equalsToken(schemaName, database.getShortName())) {
+            if (equalsToken(schemaName, database.getShortName()) || database.getIgnoreCatalogs()) {
```
`Parser.parseComment()` needs similar logic.

aschoerk replied:
The case is, I create tests for existing modules, developed by other teams. So changing of their code is the very last option. They use catalogs in their productive code, my problem is how to use H2 in more complicated tests. Usage of multiple h2-databases is also no option since I must be able to handle the accessing of different catalogs in one transaction.

aschoerk replied:
I will look into parsecomment, and handle the test-error.

aschoerk replied:
See TestIgnoreCatalogs.canCommentOn()

katzyn commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -7290,6 +7290,13 @@ private Prepared parseSet() {
+
+            database.setIgnoreCatalogs(readBooleanSetting());
+
+            return new NoOperation(session);
```
You can't change this setting directly here. Parser shouldn't execute any commands.

aschoerk replied:
Ahh, therefore you create a command. ok

aschoerk replied:
Added to SetTypes, adapted Set

katzyn commented on h2/src/test/org/h2/samples/CatalogTest.java:
```diff
@@ -0,0 +1,63 @@
+ * A very simple class that shows how to load the driver, create a database,
+ * create a table, and insert some data.
+ */
+public class CatalogTest {
```
The test case must be located in `h2/src/test` directory and be invoked from the `TestAll`.

aschoerk replied:
Ok, thanks

katzyn replied:
You need to extend and implement the `TestDb`. Take a look on existing test cases.

aschoerk replied:
created TestIgnoreCatalogs

katzyn replied:
Please, remove the `samples/CatalogTest.java`. It isn't needed.

aschoerk replied:
done

forchid commented on h2/src/test/org/h2/samples/CatalogTest.java:
```diff
@@ -0,0 +1,63 @@
+                    stat.execute("insert into catalog1.dbo.test2 values(2, 'Hello2')");
+                    throw new RuntimeException("expected exception because of not existing catalog name.");
+
+                } catch (Exception e) {
```
Here shouldn't catch "Exception", because of it's the super class of "RuntimeException".

aschoerk replied:
Ahhh, I will catch SQLException.

aschoerk replied:
done

katzyn reviewed (commented):
Thank you for your contribution!

Personally I don't like that idea. As you already wrote, it may hide real errors and its usability is questionable, people can have multiple databases with the same schemata. It would be more reasonable to add the true implementation of catalogs with automatic creation of linked tables, for example. Of course, it requires more work.

forchid commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -4861,7 +4861,7 @@ private String readIdentifierWithSchema2(String s) {
         s = readColumnIdentifier();
         if (currentTokenType == DOT) {
-            if (equalsToken(schemaName, database.getShortName())) {
+            if (equalsToken(schemaName, database.getShortName()) || database.getIgnoreCatalogs()) {
```
If a catalog name is equal to a schema name, what will happen?

katzyn replied:
Nothing special, the condition still be true. That's OK, but sides of condition could be swapped. 

The whole condition shouldn't be used by normal applications because H2 does not support catalogs, so it's efficiency is not very critical.

katzyn commented on h2/src/main/org/h2/engine/DbSettings.java:
```diff
@@ -334,6 +334,13 @@
+    /**
+     * Database setting <code>IGNORE_CATALOGS</code>
+     * (default: false).<br />
+     * Compress data when storing.
```
This description is for another setting.

aschoerk replied:
fixed

Commit:
fix catching of SQLException

Commit:
fixes because of review results
- parseComment
- catch SQLException
- create Test: TestIgnoreCatalogs started in TestAll
- fixed description in DbSettings
- put change of database-setting into command

katzyn commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -2158,7 +2158,7 @@ private boolean readIfExists(boolean ifExists) {
     }
 
-    private Prepared parseComment() {
+        private Prepared parseComment() {
```
Previous indentation was correct.

aschoerk replied:
oh sorry, will fix this

katzyn commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -2198,17 +2198,24 @@ private Prepared parseComment() {
             ArrayList<String> list = Utils.newSmallArrayList();
             do {
-                list.add(readUniqueIdentifier());
+                if (currentTokenType == DOT) {
```
Why you expect a dot here?

aschoerk replied:
comment on catalog..tablename, if the defaultschemaname geht left out

katzyn replied:
https://github.com/h2database/h2database/pull/2092#discussion_r319490143

aschoerk replied:
I searched how often this is used in our code, and it is. Since it is officially allowed in sybase and I think in mssql as well, I would like to keep it supported. I also tried this in postgres, there it is not possible.

katzyn commented on h2/src/main/org/h2/engine/Database.java:
```diff
@@ -315,6 +316,8 @@ public Database(ConnectionInfo ci, String cipher) {
                 TraceSystem.DEFAULT_TRACE_LEVEL_SYSTEM_OUT);
         this.cacheType = StringUtils.toUpperEnglish(
                 ci.removeProperty("CACHE_TYPE", Constants.CACHE_TYPE_DEFAULT));
+        this.ignoreCatalogs = ci.getProperty("IGNORE_CATALOGS",
```
You should use `getProperty(String, boolean)` here.

aschoerk replied:
Thinking too complicated

katzyn commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -4859,13 +4866,20 @@ private String readIdentifierWithSchema(String defaultSchemaName) {
-            if (equalsToken(schemaName, database.getShortName())) {
-                read();
-                schemaName = s;
+        if (readIf(DOT)) {
```
Why this method is so complicated now?

aschoerk replied:
I tried to support catalog..tablename where the schemaname gets left out

katzyn replied:
SQL Standard does not allow such chains of identifiers with missing parts in the middle. H2 should reject them too.

aschoerk replied:
I searched how often this is used in our code, and it is. Since it is officially allowed in sybase and I think in mssql as well, I would like to keep it supported. I also tried this in postgres, there it is not possible.

grandinj replied:
thank for doing the research. We prefer not to allow non-standard stuff like this, unless it is hidden behind a mode flag. Please have a look at the mode class, add a flag there, set that flag to true for MSSQL, and check that flag here

aschoerk replied:
see org.h2.test.db.TestIgnoreCatalogs.doesNotAcceptEmptySchemaWhenNotMSSQL()

katzyn replied:
This method is partially broken with default settings.

```SQL
SELECT * FROM PUBLIC..TEST;
```
throws _Table "PUBLIC" not found_. Before your changes the error message was better.

aschoerk replied:
Is a new ErrorCode necessary, or what would you prefer?

katzyn replied:
Just restore the previous behavior when the new setting / compatibility mode aren't activated.

aschoerk replied:
Should be fixed by the most recent commit.

katzyn commented on h2/src/main/org/h2/engine/DbSettings.java:
```diff
@@ -334,6 +335,13 @@
+    /**
+     * Database setting <code>IGNORE_CATALOGS</code>
+     * (default: false).<br />
+     * ignore Catalog-Prefix at object-Names.
```
Please, use normal spelling and punctuation here. Also the description should be more correct, now it looks like the catalogs aren't accepted by default, but actually they are always accepted. This setting only allows usage of any names.

aschoerk replied:
Would 
"Generally accept the all catalog prefixes at objectnames even if they are not equal to the current databasename."
be better?

katzyn replied:
Use something like
```
If set, all catalog names in identifiers are silently accepted
without comparing them with the short name of the database.
```

aschoerk replied:
If ok, I will copy that.

katzyn commented on h2/src/test/org/h2/test/TestAll.java:
```diff
@@ -929,6 +931,7 @@ private void testAdditional() {
         addTest(new TestFileLockProcess());
         addTest(new TestDefrag());
         addTest(new TestTools());
+        addTest(new TestIgnoreCatalogs());
```
It should be invoked from `test()` instead.

aschoerk replied:
done

katzyn commented on h2/src/test/org/h2/test/unit/TestIgnoreCatalogs.java:
```diff
@@ -0,0 +1,158 @@
+package org.h2.test.unit;
```
You need to add the standard copyright header (copy it from some other file).

This test case should be in the `test.db` package, I guess or in some other reasonable package.

aschoerk replied:
done

katzyn commented on h2/src/test/org/h2/test/unit/TestIgnoreCatalogs.java:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    public void canCommentOn() throws Exception {
+        try (Connection conn = getConnection("jdbc:h2:mem:testdb;IGNORE_CATALOGS=TRUE;init=drop all objects\\;create schema dbo\\;set schema dbo\\;")) {
```
You shouldn't use the complete JDBC URLs here and below. We need to test this functionality in all modes. They use different URLs.

Also you need to delete database after execution of `TestIgnoreCatalogs`, because in the most modes local files will be created.

aschoerk replied:
At the beginning I tried to test the dbsetting and saw no reasonable way.
I tried to avoid the leftover dbs by using mem. But I will look deeper into the other tests, how to handle that.

katzyn replied:
You can use `getConnection("ignoreCatalogs;IGNORE_CATALOGS=TRUE")`.

katzyn replied:
And `deleteDb("ignoreCatalogs");` before and after your test cases.

aschoerk replied:
That's not fixed yet. I need to look deeper into your request.

aschoerk replied:
fixed

katzyn commented on h2/src/test/org/h2/test/unit/TestIgnoreCatalogs.java:
```diff
@@ -0,0 +1,158 @@
+
+                stat.execute("create table test2(id int primary key, name varchar(255))");
+                // expect table already exists
+                assertThrows(42101, stat, "create table catalog2.dbo.test(id int primary key, name varchar(255))");
```
Do not use codes here and in other places, use the constants from the `ErrorCode` instead.

aschoerk replied:
fixed

Commit:
fixes because of review results
- fixed indent
- fixed usage getProperty in Database
- fixed comment in DbSettings
- fixed package and ErrorCodes of TestIgnoreCatalogs
- fixed placement of TestIgnoreCatalogs

Commit:
fixes because of review results
removed CatalogTest

Commit:
fixes because of review results
added copyrightheader

Commit:
fixes because of review results
corrected comment

Commit:
fixes because of review results
corrected connection-url

Commit:
fixes because of review results
added flag allowEmptySchemaValuesAsDefaultSchema set in Mode-MSSQLServer

Commit:
merge conflicts

aschoerk committed:
Update Mode.java

aschoerk committed:
Merge branch 'master' into master

katzyn commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -2198,17 +2198,24 @@ private Prepared parseComment() {
-                list.add(readUniqueIdentifier());
+                if (currentTokenType != DOT) {
+                    list.add(readUniqueIdentifier());
+                } else if(database.getMode().allowEmptySchemaValuesAsDefaultSchema) {
```
```SQL
CREATE TABLE TEST(ID INT);

COMMENT ON COLUMN ......PUBLIC.......TEST......ID IS 'comment';
```
is not rejected with default settings now.

aschoerk replied:
That happens in MSSQLServer mode, yes? In other modes the behaviour should not differ.

katzyn replied:
No, it happens in Regular mode. You broke both methods in the similar way due to incorrect conditional code.

katzyn replied:
`Parser` should not read and ignore unexpected tokens. For example, in this method you can't use `readIf(DOT)` when `.` is not valid for current parsing position.

aschoerk replied:
fixed see TestIgnoreCatalogs.canCommentOn()

aschoerk replied:
... ah that what missleading for me. I that readIf is invariant, if the token was not encountered. It fills the expectedlist, if the token was not found. Since DOT can only be expected in MSSQL-Case, that could be missleading.
I should have looked into the code. That concerns the other part I changed as well. So the check for .. must be done


> if (currentTokenType == DOT) {
>    read();
> }

or better separate the 
> database.getMode().allowEmptySchemaValuesAsDefaultSchema
changes from the other code.

katzyn replied:
SQL Server does not have `COMMENT ON` command. It's simpler to remove changes for `..` here. It was requested to support only the wrong catalog here for consistency, because this setting in available in Regular mode too where `COMMENT ON` is a valid command.

However, you created the problem with `..` in another method too and it isn't fixed yet.

katzyn replied:
BTW, do you really need `..`?

aschoerk replied:
Sadly sqls-server seems not to support columns on from version 7 onward. sybase does, so therefore I would leave the handling there. "Sadly" the dbms we are using is sybase.... At another place I commented, that the construct is used. But admittedly not in "comment on". 
So if you insist, I can remove all the changes i did for "comment on" but I would not recommend it.
The changes I did last should fix the differences encountered in the not mssql-case.

Commit:
fix not detected errors in: COMMENT ON COLUMN ......PUBLIC.......TEST......ID

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
fix different behaviour in not MSSQL-Case

katzyn commented on h2/src/docsrc/help/help.csv:
```diff
@@ -1482,6 +1482,18 @@ This setting can be appended to the database URL: ""jdbc:h2:test;IGNORECASE=TRUE
+multiple catalogs used by the same connections must be simulated. Caveat: If both catalogs contain schemas of the
+same name and if those schemas contain objects of the same name, this will lead to errors, when trying to manage,
+access or change these objects.
+This setting can be appended to the database URL: ""jdbc:h2:test;IGNORE_CATALOG=TRUE""
```
Please, use `jdbc:h2:./test;…` here (with `./`).

aschoerk replied:
ok, also fixed copy error below.

katzyn commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -2198,17 +2198,22 @@ private Prepared parseComment() {
             // if the db name is equal to the schema name
             ArrayList<String> list = Utils.newSmallArrayList();
             do {
+                if(database.getMode().allowEmptySchemaValuesAsDefaultSchema && readIf(DOT)) {
```
1. Add a space after `if`.

2. Does Sybase really accepts leading dot(s) here? `COMMENT ON .PUBLIC.TEST.ID IS ''`

katzyn replied:
Any comments? I can't test behavior of Sybase by myself.

aschoerk replied:
No it does not accept the leading dots. I wrote a test to show that the code does not accept this. see: canCommentOn()

katzyn replied:
`COMMENT ON COLUMN .TEST.ID IS ''` still works in MSSQLServer mode, perhaps it should be rejected too?

aschoerk replied:
was also rejected, but not at the place, where intended.

katzyn commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -4859,13 +4864,20 @@ private String readIdentifierWithSchema(String defaultSchemaName) {
             }
+        } else {
+            s = readColumnIdentifier();
+            if(currentTokenType == DOT) {
```
Space.

aschoerk replied:
replaced

katzyn commented on h2/src/main/org/h2/expression/SequenceValue.java:
```diff
@@ -5,6 +5,8 @@
 package org.h2.expression;
 
+import java.math.BigDecimal;
+
```
What is that? A merge glitch?

aschoerk replied:
That's the only explanation. Normally the import optimization would have removed that.

aschoerk replied:
Did some resolving of mergeconflicts in this file as well.

Commit:
fixed review results, concerning merge problem, if(-syntax and documentation.

aschoerk committed:
Merge branch 'master' into master

katzyn commented on h2/src/main/org/h2/expression/SequenceValue.java:
```diff
@@ -80,26 +80,26 @@ public void updateAggregate(Session session, int stage) {
+          case ExpressionVisitor.READONLY:
+              return current;
+          default:
+              throw DbException.throwInternalError("type="+visitor.getType());
```
Please, revert changes in this file. You corrupted the indentation somehow.

aschoerk replied:
Sorry, the merging using only github-browser-page .... Now I just copied your version.

katzyn replied:
Right now I see extra 4 spaces inside the whole `switch` statement.

Commit:
fixed merge problem

Commit:
fixed merge problem

katzyn commented on h2/src/main/org/h2/expression/SequenceValue.java:
```diff
@@ -108,4 +108,4 @@ public int getCost() {
     }
 
-}
+}
```
Now you lost an empty newline at the end of this file.

Not very critical, but this file is not related with your changes and shouldn't be touched at all.

aschoerk replied:
I can't imagine why the merge problems came up. I never intended to change anything at this file. I hope it is now in the original state.

katzyn commented on h2/src/test/org/h2/test/db/TestIgnoreCatalogs.java:
```diff
@@ -0,0 +1,204 @@
+    }
+
+    public void doesNotAcceptEmptySchemaWhenNotMSSQL() throws SQLException {
+        try (Connection conn = getConnection("ignoreCatalogs;IGNORE_CATALOGS=TRUE;init=drop all objects\\;create schema dbo\\;set schema dbo\\;")) {
```
Please also wrap all long lines in this file. Source code can have up to 120 characters per line in H2's sources, comments can have up to 100 characters.

katzyn replied:
Also it's better to remove `INIT` parameters from URLs and execute these commands in normal way. It will improve readability.

aschoerk replied:
fixed

Commit:
fixed line length, explicitly prevent leading DOTs

Commit:
fixed merge problems

Commit:
improved readability

katzyn commented:
Please, send a license statement as described here
https://h2database.com/html/build.html#providing_patches
to our mailing list:
https://groups.google.com/forum/#!forum/h2-database

Mailing list is partially pre-moderated; some new messages may not appear immediately.

katzyn commented:
Thank you!

aschoerk commented:
Thank you for the detailed review! And sorry for the many steps necessary.
